In [119]:
# import libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
# load the data
data = pd.read_csv('data.csv')

In [10]:
# load the data 
hotel_data = pd.read_excel('Hotel Reservation Analysis in SQL and Tableau.xlsx', sheet_name=None)


In [11]:
# to access the first sheet:
column_data = hotel_data['Summary']

In [13]:
# To access the second sheet:
data = hotel_data['hotel_bookings']

In [14]:
column_data

Unnamed: 0                             Unnamed: 1  \
0            1                                  hotel   
1            2                            is_canceled   
2            3                              lead_time   
3            4                      arrival_date_year   
4            5          text_formatarrival_date_month   
5            6               arrival_date_week_number   
6            7              arrival_date_day_of_month   
7            8                stays_in_weekend_nights   
8            9                   stays_in_week_nights   
9           10                                 adults   
10          11                               children   
11          12                                 babies   
12          13                        text_formatmeal   
13          14                     text_formatcountry   
14          15              text_formatmarket_segment   
15          16        text_formatdistribution_channel   
16          17                      is_repeated_guest   
17          18                 previous_cancellations   
18          19         previous_bookings_not_canceled   
19          20          text_formatreserved_room_type   
20          21          text_formatassigned_room_type   
21          22                        booking_changes   
22          23                text_formatdeposit_type   
23          24                       text_formatagent   
24          25                     text_formatcompany   
25          26                   days_in_waiting_list   
26          27               text_formatcustomer_type   
27          28                                    adr   
28          29            required_car_parking_spaces   
29          30              total_of_special_requests   
30          31          text_formatreservation_status   
31          32  calendar_todayreservation_status_date   

                                           Unnamed: 2  
0            Hotel (H1 = Re Hotel or H2 = City Hotel)  
1   Value indicating if the booking was canceled (...  
2   Number of days that elapsed between the enteri...  
3                                Year of arrival date  
4                               Month of arrival date  
5                Week number of year for arrival date  
6                                 Day of arrival date  
7   Number of weekend nights (Saturday or Sunday) ...  
8   Number of week nights (Monday to Friday) the g...  
9                                    Number of adults  
10                                 Number of children  
11                                   Number of babies  
12  Type of meal booked. Categories are presented ...  
13  Country of origin. Categories are represented ...  
14  Market segment designation. In categories, the...  
15  Booking distribution channel. The term “TA” me...  
16  Value indicating if the booking name was from ...  
17  Number of previous bookings that were cancelle...  
18  Number of previous bookings not cancelled by t...  
19  Code of room type reserved. Code is presented ...  
20  Code for the type of room assigned to the book...  
21  Number of changes/amendments made to the booki...  
22  Indication on if the customer made a deposit t...  
23      ID of the travel agency that made the booking  
24  ID of the company/entity that made the booking...  
25  Number of days the booking was in the waiting ...  
26  Type of booking, assuming one of four categori...  
27  Average Daily Rate as defined by dividing the ...  
28  Number of car parking spaces required by the c...  
29  Number of special requests made by the custome...  
30  Reservation last status, assuming one of three...  
31  Date at which the last status was set. This va...

In [3]:
data

hotel  is_canceled  lead_time  arrival_date_year  \
0       Resort Hotel            0        342               2015   
1       Resort Hotel            0        737               2015   
2       Resort Hotel            0          7               2015   
3       Resort Hotel            0         13               2015   
4       Resort Hotel            0         14               2015   
...              ...          ...        ...                ...   
119385    City Hotel            0         23               2017   
119386    City Hotel            0        102               2017   
119387    City Hotel            0         34               2017   
119388    City Hotel            0        109               2017   
119389    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
0                    July                        27   
1                    July                        27   
2                    July                        27   
3                    July                        27   
4                    July                        27   
...                   ...                       ...   
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  ...  deposit_type  agent company  \
0                          0       2  ...    No Deposit    NaN     NaN   
1                          0       2  ...    No Deposit    NaN     NaN   
2                          1       1  ...    No Deposit    NaN     NaN   
3                          1       1  ...    No Deposit  304.0     NaN   
4                          2       2  ...    No Deposit  240.0     NaN   
...                      ...     ...  ...           ...    ...     ...   
119385                     5       2  ...    No Deposit  394.0     NaN   
119386                     5       3  ...    No Deposit    9.0     NaN   
119387                     5       2  ...    No Deposit    9.0     NaN   
119388                     5       2  ...    No Deposit   89.0     NaN   
119389                     7       2  ...    No Deposit    9.0     NaN   

       days_in_waiting_list customer_type     adr  \
0                         0     Transient    0.00   
1                         0     Transient    0.00   
2                         0     Transient   75.00   
3                         0     Transient   75.00   
4                         0     Transient   98.00   
...                     ...           ...     ...   
119385                    0     Transient   96.14   
119386                    0     Transient  225.43   
119387                    0     Transient  157.71   
119388                    0     Transient  104.40   
119389                    0     Transient  151.20   

        required_car_parking_spaces  total_of_special_requests  \
0                                 0                          0   
1                                 0                          0   
2                                 0                          0   
3             

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

In [5]:
data.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

In [6]:
# For null values in a children column
null_children = data[data['children'].isnull()]
null_children

hotel  is_canceled  lead_time  arrival_date_year  \
40600  City Hotel            1          2               2015   
40667  City Hotel            1          1               2015   
40679  City Hotel            1          1               2015   
41160  City Hotel            1          8               2015   

      arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
40600             August                        32                          3   
40667             August                        32                          5   
40679             August                        32                          5   
41160             August                        33                         13   

       stays_in_weekend_nights  stays_in_week_nights  adults  ...  \
40600                        1                     0       2  ...   
40667                        0                     2       2  ...   
40679                        0                     2       3  ...   
41160                        2                     5       2  ...   

       deposit_type  agent company days_in_waiting_list    customer_type  \
40600    No Deposit    NaN     NaN                    0  Transient-Party   
40667    No Deposit   14.0     NaN                    0  Transient-Party   
40679    No Deposit    NaN     NaN                    0  Transient-Party   
41160    No Deposit    9.0     NaN                    0  Transient-Party   

        adr  required_car_parking_spaces  total_of_special_requests  \
40600  12.0                            0                          1   
40667  12.0                            0                          1   
40679  18.0                            0                          2   
41160  76.5                            0                          1   

       reservation_status reservation_status_date  
40600            Canceled              01/08/2015  
40667            Canceled              04/08/2015  
40679            Canceled              04/08/2015  
41160            Canceled              09/08/2015  

[4 rows x 32 columns]

In [10]:
# printing number of unique items
unique_company_count = data['country'].nunique()

print(f"Number of unique items in 'Your_Column_Name': {unique_company_count}")


Number of unique items in 'Your_Column_Name': 177


In [ ]:
## Importing data into sql using sql connectors

## Hotel reservation

In [2]:
# Import necessary libraries
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import pymysql
from sqlalchemy import create_engine

In [4]:
# loading the data
data = pd.read_csv('data.csv')

In [5]:
# display first 5 rows of data
data.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  deposit_type  \
0                        0                     0       2  ...    No Deposit   
1                        0                     0       2  ...    No Deposit   
2                        0                     1       1  ...    No Deposit   
3                        0                     1       1  ...    No Deposit   
4                        0                     2       2  ...    No Deposit   

   agent company days_in_waiting_list customer_type   adr  \
0    NaN     NaN                    0     Transient   0.0   
1    NaN     NaN                    0     Transient   0.0   
2    NaN     NaN                    0     Transient  75.0   
3  304.0     NaN                    0     Transient  75.0   
4  240.0     NaN                    0     Transient  98.0   

   required_car_parking_spaces  total_of_special_requests  reservation_status  \
0                            0                          0           Check-Out   
1                            0                          0           Check-Out   
2                            0                          0           Check-Out   
3                            0                          0           Check-Out   
4                            0                          1           Check-Out   

  reservation_status_date  
0              01/07/2015  
1              01/07/2015  
2              02/07/2015  
3              02/07/2015  
4              03/07/2015  

[5 rows x 32 columns]

In [24]:
# getting basic information
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

In [27]:
# checking for nan value
nan_columns = data.columns[data.isna().any()].tolist()
print("Columns with NaN values:", nan_columns)

Columns with NaN values: ['children', 'country', 'agent', 'company']


In [28]:
data.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

**Null Values**

* There are a total of four columns with missing values; children, country, agent and comapany.
* children column has 4 missing valies
* country column has 488 missing
* agent column has 16340 missing values,
* comapany has 112593 missing values

**Filling the null values**

In [30]:
# Fill null values in 'country' with unknown
data['country'] = data['country'].fillna('unknown')

# Fill null values in 'agent' with 0
data['agent'] = data['agent'].fillna(0,)


# Fill null values in 'company' with 0
data['company'] = data['company'].fillna(0)


# Fill null values in 'children' with 0
data['children'] = data['children'].fillna(0)


In [31]:
data.isnull().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
reserved_room_type                0
assigned_room_type                0
booking_changes                   0
deposit_type                      0
agent                             0
company                           0
days_in_waiting_list              0
customer_type                     0
adr                         

### Loading data into sql database

In [21]:
#establish connection to mysql database

In [62]:
# Create connection
conn = pymysql.connect(
    host='localhost',
    port=3306,
    user='root',
    password='9175',
    db='hotel_reservation'
)


In [63]:
# Create cursor
cursor = conn.cursor()

**Creating tables from the main table**

I have to create five different tables; reservation, guest, room, booking, status

**creating reservation table**

In [27]:
#creating column reservation_id
data.insert(0, 'reservation_id', range(500, 500 + len(data)))

In [120]:
reservation = data[['reservation_id','guest_id','hotel', 'is_canceled','lead_time','arrival_date_year','arrival_date_month','arrival_date_week_number','arrival_date_day_of_month', 'stays_in_weekend_nights','stays_in_week_nights']].copy()

In [121]:
# exporting dataframe
reservation.to_csv('reservation_information.csv', index = False)

In [122]:
# load data
reservation = pd.read_csv('reservation_information.csv')

In [123]:
reservation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 11 columns):
 #   Column                     Non-Null Count   Dtype 
---  ------                     --------------   ----- 
 0   reservation_id             119390 non-null  int64 
 1   guest_id                   119390 non-null  int64 
 2   hotel                      119390 non-null  object
 3   is_canceled                119390 non-null  int64 
 4   lead_time                  119390 non-null  int64 
 5   arrival_date_year          119390 non-null  int64 
 6   arrival_date_month         119390 non-null  object
 7   arrival_date_week_number   119390 non-null  int64 
 8   arrival_date_day_of_month  119390 non-null  int64 
 9   stays_in_weekend_nights    119390 non-null  int64 
 10  stays_in_week_nights       119390 non-null  int64 
dtypes: int64(9), object(2)
memory usage: 10.0+ MB


In [ ]:
# loading to mysql database 

In [127]:
# Load  reservation dataframe into database
for index, row in reservation.iterrows():
    sql = 'INSERT INTO reservation (reservation_id, guest_id, hotel, is_canceled, lead_time, arrival_date_year, arrival_date_month, arrival_date_week_number, arrival_date_day_of_month, stays_in_weekend_nights, stays_in_week_nights) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)'
    val = (int(row['reservation_id']),int(row['guest_id']), row['hotel'], row['is_canceled'], row['lead_time'], row['arrival_date_year'], row['arrival_date_month'], row['arrival_date_week_number'], row['arrival_date_day_of_month'] , row['stays_in_weekend_nights'], row['stays_in_week_nights'])
    cursor.execute(sql, val)


In [128]:
# Commit changes
conn.commit()

**Room Info table**

In [28]:
#creating column room_id
data.insert(0, 'room_id', range(100, 100 + len(data)))

In [92]:
# creatng room table 
room = data[['room_id', 'reservation_id','reserved_room_type','assigned_room_type','booking_changes','deposit_type','agent','company']].copy()

In [94]:
# exporting dataframe 
room.to_csv('room_information.csv', index = False)

In [95]:
# load data
room = pd.read_csv('room_information.csv')

In [98]:
# Load room dataframe into database
for index, row in room.iterrows():
    sql = 'INSERT INTO room (room_id, reservation_id, reserved_room_type, assigned_room_type, booking_changes, deposit_type, agent, company) VALUES (%s, %s, %s, %s, %s, %s, %s, %s)'
    val = (int(row['room_id']), int(row['reservation_id']), row['reserved_room_type'], row['assigned_room_type'], row['booking_changes'], row['deposit_type'], row['agent'], row['company'])
    cursor.execute(sql, val)


In [99]:
# Commit changes
conn.commit()

**Guest information table**

In [29]:
# creating column guest id
data['guest_id'] = np.arange(len(data))

This approach uses numpy.arange to create an array of integers from 0 to len(df)-1 and assigns it to a new column 'ID' in the DataFrame.

In [102]:
#creating guest table
guest = data[['guest_id', 'reservation_id', 'room_id','adults','children','babies','meal','country','market_segment','distribution_channel','is_repeated_guest','previous_cancellations','previous_bookings_not_canceled']].copy()

In [103]:
#exporting dataframe
guest.to_csv('guest_info.csv', index = False)

In [104]:
# load data
guest = pd.read_csv('guest_info.csv')

In [107]:
# Load  guest dataframe into database
for index, row in guest.iterrows():
    sql = 'INSERT INTO guest (guest_id, reservation_id, room_id, adults, children, babies, meal, country, market_segment, distribution_channel, is_repeated_guest, previous_cancellations, previous_bookings_not_canceled) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)'
    val = (int(row['guest_id']), int(row['reservation_id']), int(row['room_id']), row['adults'], row['children'], row['babies'], row['meal'], row['country'], row['market_segment'], row['distribution_channel'], row['is_repeated_guest'], row['previous_cancellations'], row['previous_bookings_not_canceled'])
    cursor.execute(sql, val)


In [108]:
# Commit changes
conn.commit()


**Booking details table**

In [30]:
# renaming index as id column
data.reset_index(inplace=True)
data.rename(columns={'index': 'booking_id'}, inplace=True)


This approach resets the index of the DataFrame and assigns a new default index. Then, it renames the default index column to 'id'.

In [109]:
# creating table
booking = data[['booking_id','guest_id', 'days_in_waiting_list','customer_type','adr','required_car_parking_spaces','total_of_special_requests']].copy()

In [110]:
# exporting dataframe
booking.to_csv('booking_details.csv', index = False)

In [111]:
# load data 
booking = pd.read_csv('booking_details.csv')

In [112]:
# Load booking dataframe into database
for index, row in booking.iterrows():
    sql = 'INSERT INTO booking (booking_id, guest_id, days_in_waiting_list, customer_type, adr, required_car_parking_spaces, total_of_special_requests) VALUES (%s, %s, %s, %s, %s, %s, %s)'
    val = (int(row['booking_id']), int(row['guest_id']), row['days_in_waiting_list'], row['customer_type'], row['adr'], row['required_car_parking_spaces'], row['total_of_special_requests'])
    cursor.execute(sql, val)


In [113]:
# Commit changes
conn.commit()


**Status table**

In [43]:
# creating id column
data.insert(0, 'status_id', range(1000, 1000 + len(data)))

In [114]:
# creating table
status = data[['status_id','reservation_id','booking_id','reservation_status','reservation_status_date']].copy()

In [115]:
# exporting table
status.to_csv('status.csv', index = False)

In [116]:
# load data
status = pd.read_csv('status.csv')

In [117]:
# Load status dataframe into database
for index, row in status.iterrows():
    sql = 'INSERT INTO status (status_id, reservation_id, booking_id, reservation_status, reservation_status_date) VALUES (%s, %s, %s, %s, %s)'
    val = (int(row['status_id']),int(row['reservation_id']), int(row['booking_id']), row['reservation_status'], row['reservation_status_date'])
    cursor.execute(sql, val)


In [118]:
# Commit changes
conn.commit()